github-flow.rst

```diff
@@ -0,0 +1,153 @@
+   głównego repozytorium
+
+Informacje w PR
+^^^^^^^^^^^^^^^
+
+Pull request powinien zawierać:
+
+-  krótką nazwę opisującą istotę rozwiązania (np. 'Implementacja autoryzacji
+   z OpenID'),
+-  tag dotyczący obszaru problemowego (np. 'uprawnienia'),
+-  informacje, jakie mogą być potrzebne osobie weryfikującej (np. 'To
+   rozwiązanie zawiera bibliotekę X, możliwe do użycia są jeszcze biblioteki
+   A i B, ale wybrane rozwiązanie ma ``<zalety>``. Dodatkowo, pytanie, czy nie
+   powinniśmy lepiej przetestować ``<obszar>``.'),
+-  [opcjonalnie] @-mention dla osób, do których osoba publikująca zwraca się
+   o review (np. X zajmuje się obszarem uprawnień i na pewno chcę, żeby tu
+   zajrzał),
+-  [opcjonalnie] tag 'WIP' (work-in-progress) dla pull requestów, które nie
+   implementują jeszcze w pełni rozwiązania, ale osoba publikująca chce o
+   nim publicznie rozmawiać i zwrócić się o pomoc.
+
+Rozmawiamy otwarcie
+-------------------
+
+Każdy pull request podlega ocenie osób uczestniczących w projekcie i
+musi przejść proces weryfikacji. Osoba proponująca rozwiązanie nie
+powinna być osobą jednocześnie je dodającą do projektu (w wyjątkowych
+sytuacjach może być konieczne szybkie zmerge'owanie zmian, ale nie jest
+to zalecana praktyka).
+
+Zalety code review to przede wszystkim dyskutowanie zmian przed ich
+wprowadzeniem (łatwiej jest coś zablokować, niż wyrzucić to z
+repozytorium) oraz jasny, angażujący wszystkich uczestników proces
+komunikacji.
+
+Code review
+~~~~~~~~~~~
+
+Proces weryfikacji powinien angażować:
+
+-  osobę publikującą,
+-  osoby zajmujące się obszarem problemowym,
+-  osobę zgłaszającą issue, jeśli pull request się do niego odnosi.
+
+Każdy pull request powinien być zweryfikowany przez dwóch innych
+członków zespołu i do jego zmerge'owania wystarczy większość głosów
+uczestników.
+
+W GitHubie w procesie review można:
+
+-  komentować pull requesta ("mam pewne uwagi, ale nie chcę blokować, bo to
+   tylko przecinki"),
+-  zablokować pull requesta poprzez ``request changes`` ("mam uwagi krytyczne
+   i nie chcę, żeby ktoś inny w zespole sprawdzał tego pull requesta, zanim
+   ich nie naprawisz"),
+-  zaakceptować poprzez ``approve`` ("uważam, że może to zostać zmerge'owane
+   do repozytorium w obecnej formie").
+
+Wprowadzanie zmian
+~~~~~~~~~~~~~~~~~~
+
+Osoba publikująca pull requesta powinna w trakcie komunikacji z zespołem
+wprowadzać ustalane zmiany na swoim branch. W tym celu powinna:
+
+-  zrobić ``interactive rebase`` w przypadku zmieniania istniejących commitów,
+-  dodać commity, o ile potrzebne są dodatkowe rozwiązania,
+-  zrobić ``git push origin +<nazwa_brancha>`` (force push) na swój remote
+   fork, co zupdate'uje pull requesta na GitHubie.
+
+Na GitHubie zostanie zachowana cała historia komunikacji.
+
+Bad practices
+-------------
+
+-  Nie należy robić force pusha na master brancha, w szczególności w
+   głównym repozytorium.
+-  Nie należy tworzyć commitów, które naprawiają wcześniejsze commity, w
```

Comment:
Mam na myśli połączenie commita naprawiającego z naprawianym.

W tym, że „PR, który nie jest WIP, zawsze powinien być propozycją gotową do merge'owania as-is”, jest wiele racji, ale przynajmniej w teorii może być to problematyczne. W przypadku opisanym wyżej można sobie np. wyobrazić, że commit ma 200 linijek, a poprawka doń 2 linijki i wtedy najwygodniej może być zawrzeć tę poprawkę w osobnym commicie, coby była ona wyraźnie widoczna dla recenzentów.

Inny przypadek: wyobraźmy sobie, że zgłaszamy pull requesta wprowadzającego wersję 1.0 programu (wraz z pewnymi zmianami). Chcemy podpisać odpowiedniego commita, ale dopiero po uzyskaniu akceptacji, przed samym merdżowaniem, coby nie krążyły dwie równoległe, podpisane wersje 1.0. (Oczywiście można powiedzieć, że wersję 1.0 powinien wyznaczać podpisany tag, a ten będzie jeden).